core/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -16,28 +16,37 @@ import zio.Task
 
 import scala.annotation.tailrec
 
-final case class ConfigSource[K, V](
-  getConfigValue: Vector[K] => PropertyTree[K, V],
-  sourceDescription: List[String]
-) { self =>
-  final def orElse(that: => ConfigSource[K, V]): ConfigSource[K, V] =
-    ConfigSource(
-      k => getConfigValue(k).getOrElse(that.getConfigValue(k)),
-      if (self.sourceDescription == that.sourceDescription) self.sourceDescription
-      else self.sourceDescription ++ that.sourceDescription
-    )
+sealed trait ConfigSource[K, V] { self =>
+  def sourceDescription: List[String]
+
+  def getConfigValue(path: List[K]): PropertyTree[K, V]
+
+  final def orElse(that: ConfigSource[K, V]): ConfigSource[K, V] = MergedConfigSource(self, that)
```

Comment:
To preserve laziness:

```suggestion
  final def orElse(that: => ConfigSource[K, V]): ConfigSource[K, V] = MergedConfigSource(() => self, () => that)
```

Reply:
done with by-name parameters + `lazy val`